DPPP/BDABuffer.h

```diff
@@ -0,0 +1,243 @@
+// modify it under the terms of the GNU General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The LOFAR software suite is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
+
+/// @file
+/// @brief Buffer holding base dependent averaged (BDA) data.
+/// @author Maik Nijhuis and Lars Krombeen
+
+#ifndef DPPP_BDABUFFER_H
+#define DPPP_BDABUFFER_H
+
+#include "../Common/Types.h"
+#include "../Common/UVector.h"
+
+#include <complex>
+#include <vector>
+
+namespace DP3 {
+  namespace DPPP {
+
+    class BDABuffer
+    {
+    public:
+      /**
+       * Parameter structure for indicating which buffer elements are enabled.
+       */
+      struct Fields {
+        /**
+         * This constructor is necessary because of bugs in gcc and clang.
+         * See https://stackoverflow.com/questions/53408962/try-to-understand-compiler-error-message-default-member-initializer-required-be
+         */
+        Fields(){}
+
+        bool data_ = true; ///< Enable/Disable visibilities.
+        bool flags_ = true; ///< Enable/Disable flags.
+        bool weights_ = true; ///< Enable/Disable weights.
+        bool full_res_flags_ = true; ///< Enable/Disable full res flags.
+      };
+
+      struct Row {
+        Row(double time,
+            double interval,
+            rownr_t row_nr,
+            std::size_t baseline_nr,
+            std::size_t n_channels,
+            std::size_t n_correlations,
+            std::complex<float>* data,
+            bool* flags,
+            float* weights,
+            bool* fullResFlags,
+            const double* uvw);
+
+        std::size_t GetDataSize() const {
+          return n_channels_ * n_correlations_;
+        }
+
+        const double time_; ///< Start time for the measurements in MJD seconds.
+        const double interval_; ///< Duration time for the measurements in seconds.
+        const rownr_t row_nr_;
+        const std::size_t baseline_nr_;
+        const std::size_t n_channels_;
+        const std::size_t n_correlations_;
+        std::complex<float>* const data_;
+        bool* const flags_;
+        float* const weights_;
+        bool* const full_res_flags_;
+        double uvw_[3];
+      };
+
+    public:
+      /**
+       * Create a new BDABuffer.
+       * @param pool_size Size of the memory pool for this buffer.
+       *                  (number of items)
+       * @param fields The fields that should be enabled in this buffer.
+       */
+      explicit BDABuffer(std::size_t pool_size, const Fields& fields = Fields());
```

Comment:
waarom explicit?

Reply:
If a constructor can be called with a single argument, using 'explicit' avoids implicit conversions. In this case, without 'explicit', a std::size_t can be implicitly converted to a BDABuffer when calling a function that expects a BDABuffer.

Reply:
We might be talking about different constructors -- I'm talking about the BDABuffer(std::size_t pool_size, const Fields& fields = Fields()) constructor. For the single-parameter constructor I agree with explicit, for this one there's no use, is there? (I think it explicitly disallows calling with initializer list, like BDABuffer b = {size,Field}, but I don't see why it is necessary to disallow that)

Reply:
Since `fields` has a default value, this constructor also accepts a single size_t parameter, so it is (also) a single-parameter constructor.